File: python/ch04/ahc001_grow.py
# python/ch04/ahc001_grow.py
"""AHC001 玩具: 企業点から矩形を素朴に拡大."""
from __future__ import annotations

import sys
from dataclasses import dataclass
from typing import List, TextIO, Tuple

GRID = 20


@dataclass
class Company:
    x: int
    y: int
    r: int


@dataclass
class Rect:
    a: int
    b: int
    c: int
    d: int

    def area(self) -> int:
        return (self.c - self.a) * (self.d - self.b)

    def copy(self) -> Rect:
        return Rect(self.a, self.b, self.c, self.d)


def read_problem(stream: TextIO) -> List[Company]:
    lines = stream.read().splitlines()
    n = int(lines[0])
    companies: List[Company] = []
    for i in range(1, n + 1):
        x, y, r = map(int, lines[i].split())
        companies.append(Company(x, y, r))
    return companies


def overlaps_positive(r1: Rect, r2: Rect) -> bool:
    ix = min(r1.c, r2.c) - max(r1.a, r2.a)
    iy = min(r1.d, r2.d) - max(r1.b, r2.b)
    return ix > 0 and iy > 0


def can_place(rect: Rect, others: List[Rect]) -> bool:
    if rect.a < 0 or rect.b < 0 or rect.c > GRID or rect.d > GRID:
        return False
    if rect.area() <= 0:
        return False
    return all(not overlaps_positive(rect, o) for o in others)


def grow_rect(company: Company, others: List[Rect]) -> Rect:
    x, y = company.x, company.y
    rect = Rect(x, y, x + 1, y + 1)
    expansions = [
        ("left", -1, 0, 0, 0),
        ("right", 0, 0, 1, 0),
        ("down", 0, -1, 0, 0),
        ("up", 0, 0, 0, 1),
    ]
    while rect.area() < company.r:
        best: Tuple[int, Rect] | None = None
        for _, da, db, dc, dd in expansions:
            nxt = Rect(rect.a + da, rect.b + db, rect.c + dc, rect.d + dd)
            if not can_place(nxt, others):
                continue
            gain = nxt.area() - rect.area()
            if best is None or gain > best[0]:
                best = (gain, nxt)
        if best is None:
            break
        rect = best[1]
    return rect


def solve(companies: List[Company]) -> List[Rect]:
    rects: List[Rect] = []
    for company in companies:
        rect = grow_rect(company, rects)
        rects.append(rect)
    return rects


def main() -> None:
    companies = read_problem(sys.stdin)
    rects = solve(companies)
    for rect in rects:
        print(rect.a, rect.b, rect.c, rect.d)


if __name__ == "__main__":
    main()

File: go/ch04/ahc001_grow.go
// go/ch04/ahc001_grow.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

const grid = 20

type Company struct {
	X, Y, R int
}

type Rect struct {
	A, B, C, D int
}

func (r Rect) area() int {
	return (r.C - r.A) * (r.D - r.B)
}

func readProblem(rd *bufio.Scanner) ([]Company, error) {
	var lines []string
	for rd.Scan() {
		lines = append(lines, rd.Text())
	}
	if err := rd.Err(); err != nil {
		return nil, err
	}
	n, _ := strconv.Atoi(lines[0])
	companies := make([]Company, n)
	for i := 0; i < n; i++ {
		p := strings.Fields(lines[i+1])
		x, _ := strconv.Atoi(p[0])
		y, _ := strconv.Atoi(p[1])
		r, _ := strconv.Atoi(p[2])
		companies[i] = Company{x, y, r}
	}
	return companies, nil
}

func overlapsPositive(r1, r2 Rect) bool {
	ix := min(r1.C, r2.C) - max(r1.A, r2.A)
	iy := min(r1.D, r2.D) - max(r1.B, r2.B)
	return ix > 0 && iy > 0
}

func canPlace(rect Rect, others []Rect) bool {
	if rect.A < 0 || rect.B < 0 || rect.C > grid || rect.D > grid {
		return false
	}
	if rect.area() <= 0 {
		return false
	}
	for _, o := range others {
		if overlapsPositive(rect, o) {
			return false
		}
	}
	return true
}

func growRect(c Company, others []Rect) Rect {
	rect := Rect{c.X, c.Y, c.X + 1, c.Y + 1}
	type delta struct{ da, db, dc, dd int }
	expansions := []delta{
		{-1, 0, 0, 0},
		{0, 0, 1, 0},
		{0, -1, 0, 0},
		{0, 0, 0, 1},
	}
	for rect.area() < c.R {
		bestGain := -1
		var best Rect
		found := false
		for _, e := range expansions {
			nxt := Rect{rect.A + e.da, rect.B + e.db, rect.C + e.dc, rect.D + e.dd}
			if !canPlace(nxt, others) {
				continue
			}
			gain := nxt.area() - rect.area()
			if !found || gain > bestGain {
				bestGain = gain
				best = nxt
				found = true
			}
		}
		if !found {
			break
		}
		rect = best
	}
	return rect
}

func solve(companies []Company) []Rect {
	rects := make([]Rect, 0, len(companies))
	for _, c := range companies {
		rects = append(rects, growRect(c, rects))
	}
	return rects
}

func main() {
	companies, err := readProblem(bufio.NewScanner(os.Stdin))
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}
	rects := solve(companies)
	for _, r := range rects {
		fmt.Println(r.A, r.B, r.C, r.D)
	}
}
